src/components/ResourceList/ResourceList.tsx

```diff
@@ -84,82 +78,98 @@ export interface ResourceListProps {
   onSortChange?(selected: string, id: string): void;
   /** Callback when selection is changed */
   onSelectionChange?(selectedItems: ResourceListSelectedItems): void;
-  /** Function to render each list item	 */
-  renderItem(item: any, id: string, index: number): React.ReactNode;
+  /** Function to render each list item   */
+  renderItem(item: ItemType, id: string, index: number): React.ReactNode;
   /** Function to customize the unique ID for each item */
-  idForItem?(item: any, index: number): string;
+  idForItem?(item: ItemType, index: number): string;
   /** Function to resolve an id from a item */
-  resolveItemId?(item: any): string;
+  resolveItemId?(item: ItemType): string;
 }
 
-type CombinedProps = ResourceListProps & WithAppProviderProps;
-
-class ResourceListInner extends React.Component<CombinedProps, State> {
-  static Item = ResourceItem;
+type ResourceListType = (<ItemType>(
+  value: ResourceListProps<ItemType>,
+) => ReactElement) & {
+  Item: typeof ResourceItem;
   // eslint-disable-next-line import/no-deprecated
-  static FilterControl = FilterControl;
+  FilterControl: typeof FilterControl;
+};
+
+export const ResourceList: ResourceListType = function ResourceList<ItemType>({
+  items,
+  filterControl,
+  resourceName: resourceNameProp,
+  promotedBulkActions,
+  bulkActions,
+  selectedItems = [],
+  selectable,
+  hasMoreItems,
+  loading,
+  showHeader = false,
+  totalItemsCount,
+  sortValue,
+  sortOptions,
+  alternateTool,
+  onSortChange,
+  onSelectionChange,
+  renderItem,
+  idForItem = (item: any, index: number) =>
```

Comment:
Could we define the function outside of the component? That way if we ever pass it into a callback or component it would change every render!
```suggestion
  idForItem = (item: ItemType, index: number) =>
```

Reply:
I couldn't get typescript to be happy with the `Object.hasOwnProperty` check. I did something a little sneaky that looks like this:
```
function defaultIdForItem<ItemType extends {id?: any}>(
  item: ItemType,
  index: number,
) {
  return Object.prototype.hasOwnProperty.call(item, 'id')
    ? item.id
    : index.toString();
}
```

Let me know what you think.